### ISSUE TITLE
Provide video recordings in mp4 format

### ISSUE THREAD

**Issue by becky-gilbert** [lookit/lookit-api#1255]
# TL;DR

WRT issue #1074, our upcoing version of EFP with RecordRTC provides files in `webm` format, which is the most common and widely-supported native file format for audio/video recording in web browsers[^1]. However, researchers have previously received their video files in `mp4` format when downloading from the Lookit/CHS platform, because our Pipe system was set up to automatically transcode files to `mp4` before storing them in S3. 

We should consider converting the `webm` video files to `mp4` (either optional or as the default) to maintain the status quo.

# Narrative

As a researcher, I have been receiving my video files in `mp4` format and would like to continue receivng the same file type. 

**Notes**: 

This issue is more important for researchers who use the video files as part of another programmatic workflow, such as automatic gaze coding. However I don't actually know what software researchers use for automated workflows and thus whether `webm` support is available alongside `mp4`. This would be helpful to know in order to understand how disruptive this file format change might be.

This issue should not affect researchers who only use the videos as part of a manual workflow (researcher opens each video file), except that the `webm` files may not open in certain applications such as Quicktime.

Also, there are many options available for researchers to convert video files to other formats themselves (though we understand that this is an inconvenience and could be a hurdle in some cases).

# Implementation

There are at least 3 options for implementation:
1. Automatically convert all video files to `mp4` immediately via AWS Lambda when they are saved to S3.
2. Convert video files to `mp4` upon every download request.
3. Use `webm` as the default format for downloads, and convert video files upon a specific request (i.e. add a 'download as mp4' button).

## Option 1: use Lambda

Pros: 
- faster download time for the researcher: video files are converted ASAP which means they are immediately available for download
- easier for me to set up: it would involve converting files via the Lambda Python code and storing them in an S3 bucket

Cons: 
- creates ~2x the S3 file storage, unless we also delete the `webm` files after conversion (which we probably shouldn't do since the webm file is the original data source)
- makes the Lambda function a little slower and thus a little more expensive

## Options 2/3: use Lookit API

Pros: 
- saves storage space: videos are only converted upon request, and can be deleted afterwards.
- makes the Lambda function a little faster/cheaper

Cons: 
- researchers would need to wait longer for downloads: we would probably need to email users when their videos are available (at least when file number/size is above some threshold)
- harder for me to set up: it would involve creating a new celery task and/or separate service, adding an email trigger, and adding/modifying video download views

---
[^1] Confusingly, `video/mp4` is listed as one of the available mime types for video recording:
https://github.com/muaz-khan/RecordRTC#configuration
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers
but this has been shown to still produce `webm` files (at least in some browsers) - see:
https://github.com/streamproc/MediaStreamRecorder/issues/117
https://github.com/collab-project/videojs-record/issues/92

More resources:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#audio_and_video_types
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Video_codecs
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#webm
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#mpeg-4_mp4

**Comment by bleonar5** [comment]
my intuition on this is that option 1 is best for the time being. I think we wouldn't want to push a big change that altered the type of files researchers were receiving, so we would want to include one of these options at least in the PR, and the second and third options seem very complex and time consuming to design/ implement. Depending on how costly the redundant storage and additional lambda step end up being in terms of memory/time, we could implement option 2 or 3 in a separate PR further down the line

**Comment by becky-gilbert** [comment]
Makes sense, thanks @bleonar5!

**Comment by becky-gilbert** [comment]
One of the complaints we're having about the webm format is that, when opened in the browser, they do not report the duration correctly and are not seekable. This is a [known bug](https://issues.chromium.org/issues/40482588) that relates to how the files are recorded/saved, and that the chromium team has labelled "won't fix". 

However, there are apparently custom JavaScript solutions for calculating the duration and adding it to the video metadata when it is saved, so that the resulting webm files are seekable and playback with duration info. For instance, the chromium bug thread mentions the [fix-webm-duration package](https://github.com/yusitnikov/fix-webm-duration) which looks to be actively maintained. If we add this fix into our recording system, it would specifically fix the duration/seeking issues that researchers are reporting. But the files would still be in webm format, so this would not other feedback/requests related to webm vs mp4 format (e.g. hard to find a video player that loads webm, difficult to convert files).
